src/ui/customBreakpointsUI.ts

```diff
@@ -125,27 +174,67 @@ class BreakpointsDataProvider implements vscode.TreeDataProvider<Breakpoint | Ca
       breakpoint.checkboxState = state;
 
       if (breakpoint instanceof Category) {
-        for (const child of breakpoint.children) {
+        for (const child of this.getChildren(breakpoint) as XHRBreakpoint[]) {
           if (child.checkboxState !== state) {
             child.checkboxState = state;
           }
         }
-      } else {
-        if (!enabled && breakpoint.parent.checked) {
-          breakpoint.parent.checkboxState = state;
-        } else if (enabled && breakpoint.parent.children.every(c => c.checked)) {
-          breakpoint.parent.checkboxState = state;
+      } else if (breakpoint instanceof Breakpoint || breakpoint instanceof XHRBreakpoint) {
+        const parent = this.getParent(breakpoint) as Category;
+        if (!enabled && parent.checked) {
+          parent.checkboxState = state;
+        } else if (enabled && parent.children.every(c => c.checked)) {
+          parent.checkboxState = state;
         }
       }
     }
 
     const ids = this.allBreakpoints.filter(b => b.checked).map(b => b.id);
+    const xhr = this.xhrBreakpoints.filter(b => b.checkboxState).map(b => b.id);
+    const xhrDisabled = this.xhrBreakpoints.filter(b => !b.checkboxState).map(b => b.id);
     for (const session of this._debugSessionTracker.getConcreteSessions()) {
       session.customRequest('setCustomBreakpoints', { ids });
+      session.customRequest('enableXHRBreakpoints', { ids: xhr });
+      session.customRequest('disableXHRBreakpoints', { ids: xhrDisabled });
     }
 
     this._onDidChangeTreeData.fire(undefined);
   }
+
+  addXHRBreakpoints(breakpoints: XHRBreakpoint[]) {
+    // filter out duplicates
+    breakpoints = breakpoints.filter(b => !this.xhrBreakpoints.map(e => e.id).includes(b.id));
+    const match = breakpoints.map(b => b.match);
+    for (const breakpoint of breakpoints) this.xhrBreakpoints.push(breakpoint);
+    for (const session of this._debugSessionTracker.getConcreteSessions())
+      session.customRequest('enableXHRBreakpoints', { ids: match });
+    // if all XHR breakpoints are removed, disable the category
+    if (this.xhrBreakpoints.length != 0) {
+      const category = this.categories.get(XHRBreakpointsCategory);
+      if (category) {
+        category.checkboxState ??= vscode.TreeItemCheckboxState.Checked;
+        category.collapsibleState ??= vscode.TreeItemCollapsibleState.Expanded;
+      }
+    }
```

Comment:
This and the block below can probably use a single reconciliation method like.

```ts
const category = this.categories.get(XHRBreakpointsCategory);
category.checkboxState = this.xhrBreakpoints.length && this.xhrBreakpoints.every(b => b.checkboxState === vscode.TreeItemCheckboxState.Checked)
  ? vscode.TreeItemCheckboxState.Checked : vscode.TreeItemCheckboxState.Unchecked;
```

...the block in `removeXHRBreakpoints` at least has a bug that if the only disabled URL pattern is removed, then the parent won't get re-checked. Doing the above scan is a little more work (as it would be done on every change) but would fix that.